File: src-tauri/src/commands/balance.rs
use crate::provider::UsageResult;

#[tauri::command]
pub async fn get_balance(base_url: String, api_key: String) -> Result<UsageResult, String> {
    crate::services::balance::get_balance(&base_url, &api_key).await
}

#[tauri::command(rename_all = "camelCase")]
pub async fn get_api_usage_balance(
    template_type: String,
    base_url: String,
    api_key: String,
    timeout_secs: Option<u64>,
) -> Result<UsageResult, String> {
    Ok(crate::services::api_balance::get_api_usage_balance(
        &template_type,
        &base_url,
        &api_key,
        timeout_secs.unwrap_or(10),
    )
    .await)
}

File: src/lib/query/balance.ts
import { useQuery } from "@tanstack/react-query";
import { balanceApi, type UsageResult } from "@/lib/api/balance";

const REFETCH_INTERVAL = 5 * 60 * 1000;

export type BalanceTemplateType =
  | "balance"
  | "sub2api"
  | "newapi"
  | "unsupported";

export const balanceKeys = {
  all: ["balance"] as const,
  provider: (
    providerId: string,
    templateType: BalanceTemplateType,
    baseUrl: string,
  ) => [...balanceKeys.all, providerId, templateType, baseUrl] as const,
};

export function useProviderBalanceQuery({
  providerId,
  templateType,
  baseUrl,
  apiKey,
  enabled,
}: {
  providerId: string;
  templateType: BalanceTemplateType;
  baseUrl: string | undefined;
  apiKey: string | undefined;
  enabled: boolean;
}) {
  const normalizedBaseUrl = baseUrl?.trim() ?? "";
  const normalizedApiKey = apiKey?.trim() ?? "";

  return useQuery<UsageResult>({
    queryKey: balanceKeys.provider(providerId, templateType, normalizedBaseUrl),
    queryFn: () => {
      if (templateType === "unsupported") {
        return Promise.resolve({
          success: false,
          error: "Unsupported balance provider",
        });
      }
      if (templateType === "sub2api" || templateType === "newapi") {
        return balanceApi.getApiUsageBalance(
          templateType,
          normalizedBaseUrl,
          normalizedApiKey,
        );
      }
      return balanceApi.getBalance(normalizedBaseUrl, normalizedApiKey);
    },
    enabled:
      enabled &&
      templateType !== "unsupported" &&
      normalizedBaseUrl.length > 0 &&
      normalizedApiKey.length > 0,
    refetchInterval: enabled ? REFETCH_INTERVAL : false,
    refetchIntervalInBackground: enabled,
    refetchOnWindowFocus: enabled,
    staleTime: REFETCH_INTERVAL,
    retry: 1,
  });
}
